base-files: handle packages alternatives when apk removes packages

map-b opened the pull request:
On commit 3010ab8 ("base-files: add update_alternatives function") was implemented the function to handle ALTERNATIVES when using APK (OPKG handle it internally) but in commit bcc6415 ("base-files: add compatibility for APK and OPKG") was only called when adding a package, so call it also when removing packages.

Fixes: bcc6415 ("base-files: add compatibility for APK and OPKG")
Fixes: https://github.com/openwrt/openwrt/issues/19090
Fixes: https://github.com/openwrt/openwrt/issues/16991

<details><summary>TESTs (ramips/mt7621)</summary>
<p>

| Before |
| ------|
```shell
root@OpenWrt:~# for app in gunzip gzip zcat; do ls -lF /bin/$app; done
lrwxrwxrwx    1 root     root             7 Jun 12 10:42 /bin/gunzip -> busybox*
lrwxrwxrwx    1 root     root             7 Jun 12 10:42 /bin/gzip -> busybox*
lrwxrwxrwx    1 root     root             7 Jun 12 10:42 /bin/zcat -> busybox*
root@OpenWrt:~#
root@OpenWrt:~# apk add gzip
(1/1) Installing gzip (1.13-r1)
gzip-1.13-r1.post-install: Executing script...
gzip-1.13-r1.post-install: add alternative: /bin/gunzip -> /usr/libexec/gunzip-gnu
gzip-1.13-r1.post-install: add alternative: /bin/gzip -> /usr/libexec/gzip-gnu
gzip-1.13-r1.post-install: add alternative: /bin/zcat -> /usr/libexec/zcat-gnu
OK: 15 MiB in 127 packages
root@OpenWrt:~# for app in gunzip gzip zcat; do ls -lF /bin/$app; done
lrwxrwxrwx    1 root     root            23 Jun 13 22:52 /bin/gunzip -> /usr/libexec/gunzip-gnu*
lrwxrwxrwx    1 root     root            21 Jun 13 22:52 /bin/gzip -> /usr/libexec/gzip-gnu*
lrwxrwxrwx    1 root     root            21 Jun 13 22:52 /bin/zcat -> /usr/libexec/zcat-gnu*
root@OpenWrt:~#
root@OpenWrt:~# apk del gzip
(1/1) Purging gzip (1.13-r1)
gzip-1.13-r1.pre-deinstall: Executing script...
OK: 15 MiB in 126 packages
root@OpenWrt:~# for app in gunzip gzip zcat; do ls -lF /bin/$app; done
lrwxrwxrwx    1 root     root            23 Jun 13 22:52 /bin/gunzip -> /usr/libexec/gunzip-gnu
lrwxrwxrwx    1 root     root            21 Jun 13 22:52 /bin/gzip -> /usr/libexec/gzip-gnu
lrwxrwxrwx    1 root     root            21 Jun 13 22:52 /bin/zcat -> /usr/libexec/zcat-gnu
root@OpenWrt:~# gzip
-ash: gzip: not found
root@OpenWrt:~#
```
```shell
root@OpenWrt:~# ls -lF /usr/bin/wget
lrwxrwxrwx    1 root     root            18 Jun 13 23:07 /usr/bin/wget -> /bin/uclient-fetch*
root@OpenWrt:~# apk add wget-ssl
(1/6) Installing libatomic1 (14.3.0-r4)
libatomic1-14.3.0-r4.post-install: Executing script...
(2/6) Installing libopenssl3 (3.5.0-r1)
libopenssl3-3.5.0-r1.post-install: Executing script...
(3/6) Installing libpcre2 (10.42-r1)
libpcre2-10.42-r1.post-install: Executing script...
(4/6) Installing libpthread (1.2.5-r4)
libpthread-1.2.5-r4.post-install: Executing script...
(5/6) Installing librt (1.2.5-r4)
librt-1.2.5-r4.post-install: Executing script...
(6/6) Installing wget-ssl (1.25.0-r1)
wget-ssl-1.25.0-r1.post-install: Executing script...
wget-ssl-1.25.0-r1.post-install: add alternative: /usr/bin/wget -> /usr/libexec/wget-ssl
OK: 20 MiB in 133 packages
root@OpenWrt:~# ls -lF /usr/bin/wget
lrwxrwxrwx    1 root     root            21 Jun 13 22:35 /usr/bin/wget -> /usr/libexec/wget-ssl*
root@OpenWrt:~#
root@OpenWrt:~# apk del wget-ssl
(1/6) Purging wget-ssl (1.25.0-r1)
wget-ssl-1.25.0-r1.pre-deinstall: Executing script...
(2/6) Purging libopenssl3 (3.5.0-r1)
libopenssl3-3.5.0-r1.pre-deinstall: Executing script...
(3/6) Purging libatomic1 (14.3.0-r4)
libatomic1-14.3.0-r4.pre-deinstall: Executing script...
(4/6) Purging libpcre2 (10.42-r1)
libpcre2-10.42-r1.pre-deinstall: Executing script...
(5/6) Purging librt (1.2.5-r4)
librt-1.2.5-r4.pre-deinstall: Executing script...
(6/6) Purging libpthread (1.2.5-r4)
libpthread-1.2.5-r4.pre-deinstall: Executing script...
OK: 15 MiB in 126 packages
root@OpenWrt:~# ls -lF /usr/bin/wget
lrwxrwxrwx    1 root     root            21 Jun 13 22:35 /usr/bin/wget -> /usr/libexec/wget-ssl
root@OpenWrt:~# wget
-ash: wget: not found
root@OpenWrt:~#
```

| After|
|--------| 
```shell
root@OpenWrt:~# for app in gunzip gzip zcat; do ls -lF /bin/$app; done
lrwxrwxrwx    1 root     root            12 Jun 13 19:26 /bin/gunzip -> /bin/busybox*
lrwxrwxrwx    1 root     root            12 Jun 13 19:26 /bin/gzip -> /bin/busybox*
lrwxrwxrwx    1 root     root            12 Jun 13 19:26 /bin/zcat -> /bin/busybox*
root@OpenWrt:~# apk add gzip
(1/1) Installing gzip (1.13-r1)
gzip-1.13-r1.post-install: Executing script...
gzip-1.13-r1.post-install: add alternative: /bin/gunzip -> /usr/libexec/gunzip-gnu
gzip-1.13-r1.post-install: add alternative: /bin/gzip -> /usr/libexec/gzip-gnu
gzip-1.13-r1.post-install: add alternative: /bin/zcat -> /usr/libexec/zcat-gnu
OK: 18 MiB in 151 packages
root@OpenWrt:~# for app in gunzip gzip zcat; do ls -lF /bin/$app; done
lrwxrwxrwx    1 root     root            23 Jun 13 21:43 /bin/gunzip -> /usr/libexec/gunzip-gnu*
lrwxrwxrwx    1 root     root            21 Jun 13 21:43 /bin/gzip -> /usr/libexec/gzip-gnu*
lrwxrwxrwx    1 root     root            21 Jun 13 21:43 /bin/zcat -> /usr/libexec/zcat-gnu*
root@OpenWrt:~# apk del gzip
(1/1) Purging gzip (1.13-r1)
gzip-1.13-r1.pre-deinstall: Executing script...
gzip-1.13-r1.pre-deinstall: add alternative: /bin/gunzip -> /bin/busybox
gzip-1.13-r1.pre-deinstall: add alternative: /bin/gzip -> /bin/busybox
gzip-1.13-r1.pre-deinstall: add alternative: /bin/zcat -> /bin/busybox
OK: 18 MiB in 150 packages
root@OpenWrt:~# for app in gunzip gzip zcat; do ls -lF /bin/$app; done
lrwxrwxrwx    1 root     root            12 Jun 13 21:44 /bin/gunzip -> /bin/busybox*
lrwxrwxrwx    1 root     root            12 Jun 13 21:44 /bin/gzip -> /bin/busybox*
lrwxrwxrwx    1 root     root            12 Jun 13 21:44 /bin/zcat -> /bin/busybox*
root@OpenWrt:~#
```
```shell
root@OpenWrt:~# ls -lF /usr/bin/wget
lrwxrwxrwx    1 root     root            18 Jun 13 17:24 /usr/bin/wget -> /bin/uclient-fetch*
root@OpenWrt:~# apk add wget-ssl
(1/6) Installing libatomic1 (14.3.0-r4)
libatomic1-14.3.0-r4.post-install: Executing script...
(2/6) Installing libopenssl3 (3.5.0-r1)
libopenssl3-3.5.0-r1.post-install: Executing script...
(3/6) Installing libpcre2 (10.42-r1)
libpcre2-10.42-r1.post-install: Executing script...
(4/6) Installing libpthread (1.2.5-r4)
libpthread-1.2.5-r4.post-install: Executing script...
(5/6) Installing librt (1.2.5-r4)
librt-1.2.5-r4.post-install: Executing script...
(6/6) Installing wget-ssl (1.25.0-r1)
wget-ssl-1.25.0-r1.post-install: Executing script...
wget-ssl-1.25.0-r1.post-install: add alternative: /usr/bin/wget -> /usr/libexec/wget-ssl
OK: 23 MiB in 156 packages
root@OpenWrt:~# ls -lF /usr/bin/wget
lrwxrwxrwx    1 root     root            21 Jun 13 21:49 /usr/bin/wget -> /usr/libexec/wget-ssl*
root@OpenWrt:~# apk del wget-ssl
(1/6) Purging wget-ssl (1.25.0-r1)
wget-ssl-1.25.0-r1.pre-deinstall: Executing script...
wget-ssl-1.25.0-r1.pre-deinstall: add alternative: /usr/bin/wget -> /bin/uclient-fetch
(2/6) Purging libopenssl3 (3.5.0-r1)
libopenssl3-3.5.0-r1.pre-deinstall: Executing script...
(3/6) Purging libatomic1 (14.3.0-r4)
libatomic1-14.3.0-r4.pre-deinstall: Executing script...
(4/6) Purging libpcre2 (10.42-r1)
libpcre2-10.42-r1.pre-deinstall: Executing script...
(5/6) Purging librt (1.2.5-r4)
librt-1.2.5-r4.pre-deinstall: Executing script...
(6/6) Purging libpthread (1.2.5-r4)
libpthread-1.2.5-r4.pre-deinstall: Executing script...
OK: 18 MiB in 150 packages
root@OpenWrt:~# ls -lF /usr/bin/wget
lrwxrwxrwx    1 root     root            18 Jun 13 21:49 /usr/bin/wget -> /bin/uclient-fetch*
root@OpenWrt:~#
```
</p>
</details>

efahl commented:
Excellent work!  I was thinking it had something to do with the missing `remove` call, but didn't dig deep enough.

Hmm, does this also fix https://github.com/openwrt/openwrt/issues/16991?  Related but maybe not the same, are there "alternatives" defined on the busybox applets???

---
Here are my test results (x86/64 main snapshot) showing the pre-deinstall script now resurrecting the proper binary:
```
$ ll /usr/bin/wget
lrwxrwxrwx    1 root     root            18 Jun 10 13:45 /usr/bin/wget -> /bin/uclient-fetch*

$ apk add wget-ssl
(1/1) Installing wget-ssl (1.25.0-r1)
wget-ssl-1.25.0-r1.post-install: Executing script...
wget-ssl-1.25.0-r1.post-install: add alternative: /usr/bin/wget -> /usr/libexec/wget-ssl
OK: 62 MiB in 294 packages

$ ll /usr/bin/wget
lrwxrwxrwx    1 root     root            21 Jun 11 05:50 /usr/bin/wget -> /usr/libexec/wget-ssl*

$ apk del wget-ssl
(1/1) Purging wget-ssl (1.25.0-r1)
wget-ssl-1.25.0-r1.pre-deinstall: Executing script...
wget-ssl-1.25.0-r1.pre-deinstall: add alternative: /usr/bin/wget -> /bin/uclient-fetch
OK: 61 MiB in 293 packages

$ ll /usr/bin/wget
lrwxrwxrwx    1 root     root            18 Jun 11 05:50 /usr/bin/wget -> /bin/uclient-fetch*
```

map-b commented:
Looks like aparcar accounted for that also, following the code carefully the `busybox` fallback will be created if there is no ALTERNATIVES found when removing a package with ALTERNATIVES defined.
https://github.com/openwrt/openwrt/blob/710ec82367c847977ba748958fedcb4a93dd9311/package/base-files/files/lib/functions.sh#L301-L302

Although I would say that the explicit way would be better?

map-b commented:
Finally, I also have been able to test all cases successfully 😄. I went almost crazy with the ADB block error (I think due to the apk-tools upgrade commit) only resolved by removing the apk cache files in /tmp/cache/ folder but I do not understand why. I thing it is because of sysupgrading keeping settings but nothing in /tmp/cache/ should survive during sysupgrading.
```
root@OpenWrt:~# apk add iperf3
(1/3) Installing libatomic1 (14.3.0-r4)
libatomic1-14.3.0-r4.post-install: Executing script...
(2/3) Installing libiperf3 (3.19-r1)
ERROR: libiperf3-3.19-r1: ADB block error
(3/3) Installing iperf3 (3.19-r1)
ERROR: iperf3-3.19-r1: ADB block error
2 errors; 15 MiB in 127 packages
root@OpenWrt:~# apk fix
(1/2) Installing libiperf3 (3.19-r1)
ERROR: libiperf3-3.19-r1: ADB block error
(2/2) Installing iperf3 (3.19-r1)
ERROR: iperf3-3.19-r1: ADB block error
2 errors; 15 MiB in 127 packages
root@OpenWrt:~# apk del iperf3
(1/1) Purging libatomic1 (14.3.0-r4)
libatomic1-14.3.0-r4.pre-deinstall: Executing script...
OK: 15 MiB in 126 packages

root@OpenWrt:/tmp# rm /tmp/cache/apk/APKINDEX.*    <----- After this, apk starts to works correctly
```
Also, I see a change, before I could download a package with wget and install it with apk, now apk says `error: uninstallable` 
```
root@OpenWrt:~# wget https://downloads.openwrt.org/snapshots/packages/mips_24kc/packages/iperf3-3.19-r1.apk
Downloading 'https://downloads.openwrt.org/snapshots/packages/mips_24kc/packages/iperf3-3.19-r1.apk'
Connecting to 151.101.130.132:443
Writing to 'iperf3-3.19-r1.apk'
iperf3-3.19-r1.apk   100% |*******************************|  2821   0:00:00 ETA
Download completed (2821 bytes)
root@OpenWrt:~# ls -l
-rw-r--r--    1 root     root          2821 Jun 12 10:12 iperf3-3.19-r1.apk
root@OpenWrt:~# apk add --allow-untrusted iperf3-3.19-r1.apk
ERROR: unable to select packages:
  iperf3-3.19-r1:
    error: uninstallable
    arch: mips_24kc
    satisfies: world[iperf3><Q1wu42UqqqR4eTtj9EsFtoMtV7FCw=]
```

map-b committed:
base-files: handle packages alternatives when apk removes packages

On commit 3010ab8 ("base-files: add update_alternatives function") was
implemented the function to handle ALTERNATIVES when using APK (OPKG
handle it internally) but in commit bcc6415 ("base-files: add
compatibility for APK and OPKG") was only called when adding a package,
so call it also when removing packages.

While we are here, check for a more specific *.alternatives files instead
of *.list, and remove redundant "filelist" variable definition.

Fixes: bcc6415 ("base-files: add compatibility for APK and OPKG")
Fixes: https://github.com/openwrt/openwrt/issues/19090
Fixes: https://github.com/openwrt/openwrt/issues/16991
Reported-and-tested-by: Eric Fahlgren <ericfahlgren@gmail.com>
Signed-off-by: Mario Andrés Pérez <mapb_@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/19093
Signed-off-by: Robert Marko <robimarko@gmail.com>

efahl commented:
Looks good to me!  

I re-ran with your last push and it's behaving as before, fixes the bug.

efahl commented:
Not sure what the problem you saw with the ADB block error was, maybe the package itself is/was broken?  My testing for the version upgrade to `apk` itself were done on a device on which I had installed both the old and updated version of `apk` so I could do side-by-side comparisons on a bunch of commands (most notably `apk list -I --full`), and everything ran clean.

I tested that same apk, seems to work now?

```
$ wget https://downloads.openwrt.org/snapshots/packages/mips_24kc/packages/iperf3-3.19-r1.apk
Downloading 'https://downloads.openwrt.org/snapshots/packages/mips_24kc/packages/iperf3-3.19-r1.apk'
Connecting to 151.101.66.132:443
Writing to 'iperf3-3.19-r1.apk'
iperf3-3.19-r1.apk   100% |*******************************|  2821   0:00:00 ETA
Download completed (2821 bytes)

$ apk adbdump iperf3-3.19-r1.apk
#%SCHEMA: 676B6370
# ADB block, size: 1184, compat: 0, ver: 0
info:
  name: iperf3
  version: 3.19-r1
  hashes: 156ec3e7c08e54ba4043d32fbe8b26098768ec4b
  description: |
    Iperf is a modern alternative for measuring TCP and UDP bandwidth performance, allowing the tuning of various parameters and characteristics.
  arch: mips_24kc
  license: BSD-3-Clause
  origin: feeds/packages/net/iperf3
  maintainer: Felix Fietkau <nbd@nbd.name>
  url: https://github.com/esnet/iperf
  installed-size: 73728
...
```

map-b commented:
> I think your PR as-is should solve the vast majority of these issues, even without putting it into the postrm phase (at least it's working for my use cases).

Even more, in the `postrm` phase there would not be the `*.alternatives` file so would have to change more things so no way.

> I tested that same apk, seems to work now?

But can you install that?
```
root@OpenWrt:~# apk add --allow-untrusted iperf3-3.19-r1.apk
ERROR: unable to select packages:
  iperf3-3.19-r1:
    error: uninstallable
    arch: mips_24kc
    satisfies: world[iperf3><Q1FW7D58COVLpAQ9MvvosmCYdo7Es=]
root@OpenWrt:~# apk adbdump iperf3-3.19-r1.apk
#%SCHEMA: 676B6370
# ADB block, size: 1184, compat: 0, ver: 0
info:
  name: iperf3
  version: 3.19-r1
  hashes: 156ec3e7c08e54ba4043d32fbe8b26098768ec4b
...
```

map-b commented:
Just added some test results and the Reported-and-tested-by: tag, this should be ready.

efahl commented:
Great, fixes another little nuisance bug in `apk`!

@aparcar @Ansuel
